core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,396 @@
+  private volatile X509Certificate[] trustCerts;
+  // The key store that is used to hold the trust certificates. It will always keep synced with
+  // trustCerts.
+  private volatile  KeyStore ks;
+
+  private AdvancedTlsX509TrustManager(Verification verification,  PeerVerifier peerVerifier,
+      SslSocketAndEnginePeerVerifier socketAndEnginePeerVerifier) throws CertificateException {
+    this.verification = verification;
+    this.peerVerifier = peerVerifier;
+    this.socketAndEnginePeerVerifier = socketAndEnginePeerVerifier;
+    try {
+      ks = KeyStore.getInstance(KeyStore.getDefaultType());
+      ks.load(null, null);
+    } catch (KeyStoreException | IOException | NoSuchAlgorithmException | CertificateException e) {
+      throw new CertificateException("Failed to create KeyStore", e);
+    }
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, false);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain,  String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, true);
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return trustCerts;
+  }
+
+  // If this is set, we will use the default trust certificates stored on user's local system.
+  // After this is used, functions that will update this.trustCerts(e.g. updateTrustCredentials(),
+  // updateTrustCredentialsFromFile()) should not be called.
+  public void useSystemDefaultTrustCerts() {
+    this.trustCerts = new X509Certificate[0];
+    this.ks = null;
+  }
+
+  /**
+   * Updates the current cached trust certificates as well as the key store.
+   *
+   * @param trustCerts  the trust certificates that are going to be used
+   */
+  public void updateTrustCredentials(X509Certificate[] trustCerts) throws KeyStoreException,
+      CertificateException {
+    this.trustCerts = Arrays.copyOf(trustCerts, trustCerts.length);
+    // Clear the key store by re-creating it.
+    try {
+      ks = KeyStore.getInstance(KeyStore.getDefaultType());
```

Comment:
Sounds good! I've made the changes to fully initialize a key store and then swap it with the existing one at the end.

The trustCerts is indeed unnecessary, and the reason for keeping it(and also the key and certs on the key manager side) was to reduce some code...Getting these contents back from the key store requires a lot of type casting/converting.

But on a secondary thought, the logic to have two fields in sync seemed adding unnecessary burden while understanding the code.  I've changed the code to merely based on the key store. That would require the following changes:
1. add additional "password" field for the key manager, since the key store wouldn't allow me to store that without a password
2. add additional "aliases" field for the trust manager, since we will need to keep track of all the alias for the trust certs, when later getting them from `getAcceptedIssuers`

I don't have strong opinions on these two ways. Feel free to let me know if we prefer some other ways. Thank you so much!